### PR TITLE
libkeccak 1.2 (new formula)

### PR DESCRIPTION
**Pull request by zbeekman**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/libkeccak.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    system "make", "install", "OSCONFIGFILE=macos.mk", "PREFIX=#{prefix}"
+  end
+
+  test do
+    (testpath/"test.c").write <<~EOS
+      #include <libkeccak.h>
+      int main() {
+      libkeccak_spec_t spec;
+        libkeccak_state_t state;
+        return 0;
```
Comment (zbeekman) [inline Formula/libkeccak.rb]: This test probably needs to be improved. It's not even linking to the library. I just wanted to open this PR and see if folks from PR #32618 were interested in moving this forwards now that upstream patched the build.

CC: @danteu @nremond @cleishm

**Review by zbeekman — COMMENTED** [review]
Needs better test

**Comment by zbeekman** [comment]
This needs a better test before it can merge. Will try to get to it this weekend.

**Comment by zbeekman** [comment]
> * GitHub repository not notable enough (<30 forks, <30 watchers and <75 stars)

The idea is to get this in, so that https://github.com/maandree/sha3sum can be added, which does meet the metrics. This could be vendored fairly easily in sha3sum... what do other @Homebrew/core people think?

Also, it's been a while since I submitted a new formula; I was surprised when audit didn't complain, but I guess the bot has taken over that job.

**Comment by fxcoudert** [comment]
> so that https://github.com/maandree/sha3sum can be added, which does meet the metrics

Like, barely. And is not exactly thriving with activity.

**Comment by sjackman** [comment]
> Also, it's been a while since I submitted a new formula; I was surprised when audit didn't complain, but I guess the bot has taken over that job.

That's surprising to me too. I'd expect CI to fail with 🔴 if `brew audit` is not satisfied. Is it behaving as desired/designed?

**Comment by sjackman** [comment]
`coreutils` has `sha1sum sha224sum sha256sum sha384sum sha512sum`. Can `coreutils` be used rather than `maandree/sha3sum`?

**Comment by sjackman** [comment]
I suspect that `openssl` also has implementations of these SHA algorithms.

**Comment by sjackman** [comment]
```console
$ openssl dgst -? 2>&1 | grep sha
-sha            to use the sha message digest algorithm
-sha1           to use the sha1 message digest algorithm
-sha224         to use the sha224 message digest algorithm
-sha256         to use the sha256 message digest algorithm
-sha384         to use the sha384 message digest algorithm
-sha512         to use the sha512 message digest algorithm
```

**Comment by zbeekman** [comment]
@sjackman wrote:
> I suspect that `openssl` also has implementations of these SHA algorithms.

If new enough, yes, OpenSSL at least claims support for SHA-3 as of version 1.1.1 released in September 2018. OpenSSL shipped by homebrew looks to be `1.0.2q`. Furthermore, `libkeccak` and `sha3sum` are more than just providing SHA3 functionality, there are different variants of SHA-3, and closely related hashing algorithms like SHAKE and Kangaroo12.

Also, there is some drama around the parameters that NIST selected for SHA3, relative to the winning teams suggestions. IMO, more implementations are better for the community.

I personally think this is a nice, small, library written in strict, standard conforming C, without any dependencies. It may not be active, but it doesn't need to be active; it's a standardized implementation.

[From wikipedia](https://en.wikipedia.org/wiki/SHA-3):

>SHA-3 is a subset of the broader cryptographic primitive family Keccak (/ˈkɛtʃæk, -ɑːk/), designed by Guido Bertoni, Joan Daemen, Michaël Peeters, and Gilles Van Assche, building upon RadioGatún. Keccak's authors have proposed additional uses for the function, not (yet) standardized by NIST, including a stream cipher, an authenticated encryption system, a "tree" hashing scheme for faster hashing on certain architectures, and AEAD ciphers Keyak and Ketje.

From sha3sum repo:

```
NAME
	sha3sum - Keccak, SHA-3, SHAKE, and RawSHAKE checksum utilities

SYNOPSIS
	keccaksum      [option ...] [file ...]
	keccak-224sum  [option ...] [file ...]
	keccak-256sum  [option ...] [file ...]
	keccak-384sum  [option ...] [file ...]
	keccak-512sum  [option ...] [file ...]
	sha3-224sum    [option ...] [file ...]
	sha3-256sum    [option ...] [file ...]
	sha3-384sum    [option ...] [file ...]
	sha3-512sum    [option ...] [file ...]
	shake256sum    [option ...] [file ...]
	shake512sum    [option ...] [file ...]
	rawshake256sum [option ...] [file ...]
	rawshake512sum [option ...] [file ...]

DESCRIPTION
	Fully configurable byte-orientated checksum utilities
	for Keccak and its close derivatives SHA-3, SHAKE and
	RawSHAKE.

	This utilities can generate checksums or verify the
	checksums of files.
```

**Comment by zbeekman** [comment]
@fxcoudert said (about sha3sum):
> Like, barely. And is not exactly thriving with activity.

Well it's written in standard conforming C, with a single dependency, implementing standardized crypto primitives and hash functions. So the sleepiness of the repo may be an indication of stability, which makes our lives easier. (Unless of course users play with it and find lots of bugs...)

I'm a fan of adding these two packages. The code is simple, implements standardized crypto functions, and has minimal/no dependencies. Same with the build system. If we only want to have a single formula, we could vendor `libkeccak` in `sha3sum`... but the formulas are so easy/simple I think that would create more work/confusion than needed.

I feel like this would be a nice addition, because, AFAICT, there are no other SHA3 or keccak libraries or utilities in Homebrew. I don't anticipate people rushing to adopt SHA3 over SHA2, but when SHA2 breaks, the more time people have had to experiment with and poke SHA3 the better.

At any rate, I will cede veto power to @fxcoudert and other seasoned core maintainers if they have strong objections.

**Comment by sjackman** [comment]
I have no objection. I only wanted to point out alternative implementations.

**Comment by zbeekman** [comment]
> I have no objection. I only wanted to point out alternative implementations.

As far as I can tell, however (maybe my other comment before the last one got burried), there *are no other implementations available __right now__ through Homebrew*.

**Comment by sjackman** [comment]
Homebrew has `openssl@1.1` 1.1.1a.

**Comment by zbeekman** [comment]
Ahhhh I missed that detail. Fair enough.

I’ll have to see which SHA3 they ship.

**Comment by sjackman** [comment]
```console
$ /usr/local/opt/openssl@1.1/bin/openssl help 2>&1 | grep sha3
sha224            sha256            sha3-224          sha3-256
sha3-384          sha3-512          sha384            sha512
```

**Comment by zbeekman** [comment]
Ha, thanks, I can look it up when I’m back too. There are some others too
(e.g., SHAKE)

**Comment by MikeMcQuaid** [comment]
> The idea is to get this in, so that https://github.com/maandree/sha3sum can be added, which does meet the metrics. This could be vendored fairly easily in sha3sum... what do other `@Homebrew/core` people think?

I was initially leaning towards vendoring but this package is super simple to build and seems useful so I can't see it failing on us so I'm 👍 in adding it as a separate package.

Note the metrics are there to make it easy for us to say no rather than hard for us to say yes.

> I personally think this is a nice, small, library written in strict, standard conforming C, without any dependencies. It may not be active, but it doesn't need to be active; it's a standardized implementation.
> I'm a fan of adding these two packages.

Additionally on the metrics above: if a currently active maintainer is in favour of adding a package and the only audit violations are the popularity metrics I'm 👍 on always letting it through.

**Comment by zbeekman** [comment]
@MikeMcQuaid Thanks for weighing in! I just want to make sure that I also stay on @fxcoudert's good side, too. 😉

**Comment by fxcoudert** [comment]
Oh I'm definitely not vetoing any maintainer's PR… 😆

**Pull request closed without merging**